Better support for @wordpress/element Fragment JSX

sirreal opened the pull request:
See https://github.com/WordPress/gutenberg/pull/15120

Matches Gutenberg babel preset for JSX:
https://github.com/WordPress/gutenberg/blob/af6e5ecd8b398e08127ad1b3c784ce97899de100/packages/babel-preset-default/index.js#L58-L70

Provides better `calypso-build` support for `<></>`.

#### Testing instructions

* Extensions and apps built with `calypso-build` should continue to work as before.

Provided that `<></>` syntax was basically unsupported by the previous setup, this should introduce few or no changes for consumers.

It will allow `<Fragment></Fragment>` to be replaced by `<></>`, leveraging the correct `@wordpress/element` Fragment.

sirreal committed:
Better support for @wordpress/element Fragment JSX

See https://github.com/WordPress/gutenberg/pull/15120

ockham reviewed (approved):
Makes a ton of sense :ship:  

Do we need a Changelog entry for this?

sirreal committed:
Upgrade dependency

sirreal committed:
Add CHANGELOG entry

sirreal commented:
A bit of detail, you can test this by replacing an app entry point with something including a `<></>` Fragment and building it (`npx lerna run build --scope='*/o2-blocks'`).

```patch
diff --git a/apps/o2-blocks/src/editor.js b/apps/o2-blocks/src/editor.js
index 1cbc25a62b..b1ddef73ad 100644
--- a/apps/o2-blocks/src/editor.js
+++ b/apps/o2-blocks/src/editor.js
@@ -1,7 +1,3 @@
-/**
- * Internal dependencies
- */
-import './category';
-import './editor-notes/editor';
-import './prev-next/editor';
-import './todo/editor';
+function MyComponent() {
+	return <>👋 Howdy!</>;
+}
```

This results in a failed build on master `transform-react-jsx: pragma has been set but pragmafrag has not been set`.

Actually testing this with `<></>` led me to discover that the dependency responsible for this transformation was never updated so this wasn't actually working properly. I've updated the necessary dependency and added a changelog entry. Now when we try our test build we get the following output:

```js
/***/ "./src/editor.js":
/*!***********************!*\
  !*** ./src/editor.js ***!
  \***********************/
/*! no exports provided */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _wordpress_element__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! @wordpress/element */ "@wordpress/element");
/* harmony import */ var _wordpress_element__WEBPACK_IMPORTED_MODULE_0___default = /*#__PURE__*/__webpack_require__.n(_wordpress_element__WEBPACK_IMPORTED_MODULE_0__);


function MyComponent() {
  return Object(_wordpress_element__WEBPACK_IMPORTED_MODULE_0__["createElement"])(_wordpress_element__WEBPACK_IMPORTED_MODULE_0__["Fragment"], null, "\uD83D\uDC4B Howdy!");
}

/***/ }),

/***/ "@wordpress/element":
/*!*****************************!*\
  !*** external "wp.element" ***!
  \*****************************/
/*! no static exports found */
/***/ (function(module, exports) {

module.exports = wp.element;

/***/ })
```

Squint and you can see that `createElement` and `Fragment` are properties of `wp.element`.

<details>
<summary>Full output</summary>

```js
(function(e, a) { for(var i in a) e[i] = a[i]; }(window, /******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, { enumerable: true, get: getter });
/******/ 		}
/******/ 	};
/******/
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = function(exports) {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/
/******/ 	// create a fake namespace object
/******/ 	// mode & 1: value is a module id, require it
/******/ 	// mode & 2: merge all properties of value into the ns
/******/ 	// mode & 4: return value when already ns object
/******/ 	// mode & 8|1: behave like require
/******/ 	__webpack_require__.t = function(value, mode) {
/******/ 		if(mode & 1) value = __webpack_require__(value);
/******/ 		if(mode & 8) return value;
/******/ 		if((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;
/******/ 		var ns = Object.create(null);
/******/ 		__webpack_require__.r(ns);
/******/ 		Object.defineProperty(ns, 'default', { enumerable: true, value: value });
/******/ 		if(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));
/******/ 		return ns;
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";
/******/
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = "./src/editor.js");
/******/ })
/************************************************************************/
/******/ ({

/***/ "./src/editor.js":
/*!***********************!*\
  !*** ./src/editor.js ***!
  \***********************/
/*! no exports provided */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _wordpress_element__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! @wordpress/element */ "@wordpress/element");
/* harmony import */ var _wordpress_element__WEBPACK_IMPORTED_MODULE_0___default = /*#__PURE__*/__webpack_require__.n(_wordpress_element__WEBPACK_IMPORTED_MODULE_0__);


function MyComponent() {
  return Object(_wordpress_element__WEBPACK_IMPORTED_MODULE_0__["createElement"])(_wordpress_element__WEBPACK_IMPORTED_MODULE_0__["Fragment"], null, "\uD83D\uDC4B Howdy!");
}

/***/ }),

/***/ "@wordpress/element":
/*!*****************************!*\
  !*** external "wp.element" ***!
  \*****************************/
/*! no static exports found */
/***/ (function(module, exports) {

module.exports = wp.element;

/***/ })

/******/ })));
//# sourceMappingURL=editor.js.map
```

</details>

ockham commented:
Thanks :grimacing: Sorry for not testing with `</>` myself earlier :sweat_smile: 

Tree-shaking seems to outsmart me, so I have to actually make webpack believe the file does something in order to repro on `master`:

```diff
diff --git a/apps/o2-blocks/src/editor.js b/apps/o2-blocks/src/editor.js
index 1cbc25a62b..e5fe650697 100644
--- a/apps/o2-blocks/src/editor.js
+++ b/apps/o2-blocks/src/editor.js
@@ -1,7 +1,3 @@
-/**
- * Internal dependencies
- */
-import './category';
-import './editor-notes/editor';
-import './prev-next/editor';
-import './todo/editor';
+export default function MyComponent() {
+       return <>👋 Howdy!</>;
+}

```

ockham commented:
Merge once e2e tests pass :+1:

ockham reviewed (dismissed):
Merge once e2e tests pass +1

sirreal commented:
There may be some valid failures on e2e tests I'll need to take a look at.

sirreal commented:
After a rebase and a fresh deps update, there aren't any issues with tests and I think we can move forward with this.

I'd appreciate reviews, especially from @Automattic/cylon @Automattic/serenity and @Automattic/ajax who are using the build tool in the monorepo.

sirreal commented:
There are indeed failing e2e tests when accessing the editor, apparently from `process.env.FORCE_REDUCED_MOTION`:

```
entry-main.7e542f228a4b55baac16.min.js:formatted:1180 ReferenceError: process is not defined
    at Object../node_modules/@wordpress/blocks/build-module/index.js (199.2cb57698dc1bf19debcc.min.js:formatted:1114)
    at n (simplepaymemts2m7cn2m7.wordpress.com?retry=1:34)
    at Module../client/gutenberg/editor/index.js (gutenberg-editor.c502fffddc3f0da7a380.min.js:1)
    at n (simplepaymemts2m7cn2m7.wordpress.com?retry=1:34)
    at /block-editor/post/async https:/hash-c174c1d5ea4a7dae21d7e8a61969f12884f50901.calypso.live/calypso/evergreen/entry-main.7e542f228a4b55baac16.min.js:1
    at /block-editor/post/async https:/hash-c174c1d5ea4a7dae21d7e8a61969f12884f50901.calypso.live/calypso/evergreen/entry-main.7e542f228a4b55baac16.min.js:1
```

A quick search suggests this is coming from unguarded `process` access in Gutenberg: https://github.com/WordPress/gutenberg/blob/eba2fc8ba1ffbd5997cf47b3aa7002f8a5597e9f/packages/compose/src/hooks/use-reduced-motion/index.js#L19

This could easily come from some of the @wordpress package updates in shrinkwrap, I'll need to dig around and see if issues have been reported and find out exactly what's going on.

sirreal commented:
I believe failures in the editor e2e tests after running `update-deps` were caused by this issue:

https://github.com/WordPress/gutenberg/pull/16679

We'll need to wait for that to be fixed and released before upgrading `@wordpress/compose` (cc: @Automattic/team-calypso / #32607)

sirreal commented:
I'm not excited about the fix in df706bc - we seem to have `node_modules` installed under packages, which means that from `apps` we can't resolve the dependency under `packages/calypso-build`, which is confusing. I didn't expect `node_modules` to be created under package directories 😕 .

`npm dedupe` may be a solution, but I'd like to fully understand the issue before settling on a fix.

sirreal committed:
Add necessary process.env to DefinePlugin

sirreal committed:
Update deps

sirreal commented on packages/calypso-build/webpack.config.js:
```diff
@@ -121,6 +121,9 @@ function getWebpackConfig(
 			new webpack.DefinePlugin( {
 				'process.env.NODE_ENV': JSON.stringify( process.env.NODE_ENV ),
+				'process.env.FORCE_REDUCED_MOTION': JSON.stringify(
+					!! process.env.FORCE_REDUCED_MOTION || false
```
`@wordpress/compose` requires this (or `node: true`) in webpack config.

https://github.com/WordPress/gutenberg/pull/16679

sirreal committed:
Add FORCE_REDUCED_MOTION replacement to base webpack

sirreal commented:
This seems to be working well now with the `process.env.FORCE_REDUCED_MOTION` changes. Reviews appreciated.

ockham commented:
Interesting that desktop e2e tests are passing now -- only mobile ones are failing...

ockham commented:
> Interesting that desktop e2e tests are passing now -- only mobile ones are failing...

Ooh, rebuilding seems to have fixed them :tada:

sirreal commented:
Rebased to fix conflict, but noting full e2e was passing:

- https://circleci.com/gh/Automattic/wp-e2e-tests-for-branches/20810
- https://circleci.com/gh/Automattic/wp-e2e-tests-for-branches/20909

ockham reviewed (approved):
This appears to be working fine now! :shipit: 

(I might've noticed that build time has gone up :confused: )

sirreal commented:
Full e2e desktop passing: https://circleci.com/gh/Automattic/wp-e2e-tests-for-branches/21057

ockham commented:
@sirreal FWIW, I've stopped using the 'Needs e2e Testing' label to retrigger e2e tests if just one of them is failing. Instead, I rebuild manually from within CircleCI's admin dashboard.
